File: Leetcode/Average of bs levels/average_of_bs_levels.cpp
class Solution {
public:
    vector<double> averageOfLevels(TreeNode* root) {
        vector<double> ans;
        queue<TreeNode*> q;
        q.push(root);
        while(!q.empty()){
            long temp=0;
            int size= q.size();
            for(int i=0;i<size;i++){
                TreeNode * node = q.front();
                q.pop();
                if(node->left!=NULL) q.push(node->left);
                if(node->right!=NULL) q.push(node->right);
                temp+=node->val;
            }
            ans.push_back((double)temp/size);
        }
        return ans;
    }
};
// Time Complexity : O(N), where N is the number of nodes in the given Tree.
// Space Complexity : O(M), where M is the maximum number of nodes at any level in the binary tree.

File: Leetcode/Average of bs levels/average_of_bs_levels.java
class Solution {
    public List<Double> averageOfLevels(TreeNode root) {
        Queue<TreeNode> queue = new LinkedList<>();
        ArrayList<Double> list = new ArrayList<>();
        
        queue.add(root);
        
        while(!queue.isEmpty()){
            double size = (double)queue.size();
            double average = 0;
            
            for(int i = 0; i<size; i++){
                TreeNode node = queue.poll();
                average += (double)node.val;
                
                if(node.left!=null) queue.add(node.left);
                if(node.right!=null) queue.add(node.right);
            }
            list.add(average/size);
        }
        
        return list;
    }
}